File: 103. Binary Tree Zigzag Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        queue, temp, res, rev = [root], [], [], True
        while queue:
            r = []
            for node in queue:
                r.append(node.val)
                if node.left:
                    temp.append(node.left) 
                if node.right:
                    temp.append(node.right)
            
            res.append(r if rev else r[::-1])
            rev = not rev
            queue = temp
            temp = []
        return res

File: 103. Binary Tree Zigzag Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null) {
            return ans;
        }
        Deque<TreeNode> q = new ArrayDeque<>();
        q.add(root);
        boolean rev = false;
        while (!q.isEmpty()) {
            List<Integer> temp = new ArrayList<>();
            Deque<TreeNode> next = new ArrayDeque<>();
            for (TreeNode node : q) {
                temp.add(node.val);
                if (node.left != null) {
                    next.add(node.left);
                }
                if (node.right != null) {
                    next.add(node.right);
                }
            }
            if (rev) {
                Collections.reverse(temp);
            }
            rev = !rev;
            ans.add(temp);
            q = next;
        }
        return ans;
    }
}